File: docker/Dockerfile.java
# -------------------------------------------
#  Java 应用 - 多阶段 Dockerfile
#  使用方式: docker build --build-arg JDK_VERSION=17 -f docker/Dockerfile.java .
# -------------------------------------------

# -- 构建阶段 --
ARG JDK_VERSION=17
FROM maven:3.9-eclipse-temurin-${JDK_VERSION} AS builder
WORKDIR /build
COPY pom.xml .
# 利用 Docker 缓存层: 先下载依赖，再拷贝源码
RUN mvn dependency:go-offline -B
COPY src ./src
RUN mvn clean package -DskipTests -B

# -- 运行阶段 --
FROM eclipse-temurin:${JDK_VERSION}-jre-alpine
ARG JDK_VERSION=17
WORKDIR /app
RUN addgroup -S appgroup && adduser -S appuser -G appgroup
COPY --from=builder /build/target/*.jar app.jar
USER appuser
EXPOSE 8080
ENTRYPOINT ["java", "-jar", "app.jar"]

File: docker/Dockerfile.go
# -------------------------------------------
#  Go 应用 - 多阶段 Dockerfile
#  使用方式: docker build --build-arg GO_VERSION=1.22 -f docker/Dockerfile.go .
# -------------------------------------------

# -- 构建阶段 --
FROM golang:${GO_VERSION}-alpine AS builder
ARG GO_VERSION=1.22
WORKDIR /build
RUN apk add --no-cache ca-certificates tzdata
COPY go.mod go.sum ./
RUN go mod download
COPY . .
RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build \
    -ldflags="-s -w" \
    -o app .

# -- 运行阶段 --
FROM scratch
ARG GO_VERSION=1.22
COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
COPY --from=builder /usr/share/zoneinfo /usr/share/zoneinfo
COPY --from=builder /build/app /app
EXPOSE 8080
ENTRYPOINT ["/app"]
